src/lib/snark_params/pedersen.ml

```diff
@@ -100,10 +126,72 @@ end) : S with type curve := Curve.t and type Digest.t = Field.t = struct
       let x, _y = Curve.to_coords t.acc in
       x
 
-    let salt params s = update_fold (create params) (Fold.string_triples s)
+    let salt params chunk_table s =
+      update_fold (create params chunk_table) (Fold.string_triples s)
   end
 
-  let hash_fold s fold = State.update_fold s fold
+  (* break triples list into list of chunks and maybe a less-than-full chunk *)
+  let chunk_triples (triples : bool Triple.t list) :
+      Chunk.t list * Chunk.t option =
+    let rec loop triples accum =
+      if List.is_empty triples then
+        (* only full chunks, no straggler *)
+        (List.rev accum, None)
+      else
+        let some, rest = List.split_n triples Chunk.size in
+        if List.is_empty rest && List.length some < Chunk.size then
+          (* less than full chunk *)
+          (List.rev accum, Some some)
+        else loop rest (some :: accum)
+    in
+    loop triples []
+
+  (* curve point from a list of chunks (and possible straggler) *)
+  let update_from_chunks (s : State.t)
+      ((chunks, maybe_chunk) : Chunk.t list * Chunk.t option) : State.t =
+    let curve_points_table = s.chunk_table.curve_points_table in
+    let triples_consumed = s.triples_consumed in
+    (* if we have consumed a number of triples that is not an integral number of chunks,
+       can't use the chunk table; 
+    *)
+    if not (Int.equal (triples_consumed mod Chunk.size) 0) then
+      failwith
+        "update_from_chunks: can't use chunk table given the number of \
+         consumed triples" ;
+    let param_offset = s.triples_consumed / Chunk.size in
+    let get_chunk_state i accum chunk =
+      let n = Chunk.to_int chunk in
+      let g = curve_points_table.(param_offset + i).(n) in
+      { accum with
+        State.triples_consumed=
+          accum.State.triples_consumed + List.length chunk
+      ; State.acc= Curve.add accum.State.acc g }
+    in
+    let state = List.foldi chunks ~init:s ~f:get_chunk_state in
+    (* handle less-than-full straggler chunk, if needed *)
+    match maybe_chunk with
+    | None -> state
+    | Some chunk -> State.update_fold state (Fold.of_list chunk)
+
+  (* if the number of triples consumed isn't on a chunk boundary,
+     process just enough triples to get to a chunk boundary
+   *)
+  let synch_to_chunk_boundary (s : State.t) (fold : bool Triple.t Fold.t) :
```

Comment:
it's probably not worth doing because I assume allocating this list is a pretty small part of the overall computation, but we *could* do all this synching and so on in a streaming manner within the fold. Like basically have

```ocaml
type state =
  | Syncing of { acc : Curve.t; triples_consumed : int }
  | Accumulating_chunk of { acc; chunk_index_acc : int; triples_consumed : int; triples_consumed_intra_chunk : int; chunk_so_far_rev : bool Triple.t list }

let update : state -> bool Triple.t -> state = ...
```

Reply:
Yes, I've thought of a way to do the synching as part of the fold, a bit different than your suggestion.